### PR TITLE
Code generation utility in vscode ext

### PR DESCRIPTION
**Pull request by dsgouda**
Moving from https://github.com/Azure/openapi-lint-extension/pull/68

### CODE REVIEW

**Comment by dsgouda** [comment]
@fearthecowboy @olydis @veronicagg PTAL

File: src/server/document-context.ts [hunk 1]
```diff
@@ -34,6 +38,35 @@ export class DocumentContext extends EventEmitter implements IFileSystem {
...
+    codeGeneratorInstance.AddConfiguration({ "verbose": true });
+    codeGeneratorInstance.AddConfiguration(additionalConfig);
+
+    const opDirName: string = (this.configurationFile && fs.existsSync(this.configurationFile)) ?
+      path.basename(path.dirname(this.configurationFile)) : 'out';
+
+    // where to output the generated code (clear the directory before generating code)
+    const opPath: string = path.join(os.tmpdir(), opDirName);
+    codeGeneratorInstance.AddConfiguration({ "output-folder": opPath });
+    result['outputFolder'] = opPath;
```
Comment (olydis) [inline src/server/document-context.ts]: could you type the `result` object (and all the places it flows)?

File: src/client/splitpane.ts [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    if (args) {
+      opfname = args.join('.') + '.' + opfname;
+    }
+
+    const filePath: string = path.normalize(path.join(os.tmpdir(), opfname));
+    let content: string = '## Generated Code\n';
+    const generatedFiles = resultObj['generatedFiles'];
+
+    Object.keys(generatedFiles).forEach(filePath => {
+      content += '### File Name:'
```
Comment (olydis) [inline src/client/splitpane.ts]: why not just use the filename as the heading to the file content? E.g.

## [`Models/Cowbell.cs`](%TMP%\Models\Cowbell.cs)

``` C#
// stuff
class Cowbell
{
    // stuff
}
```

Comment (olydis) [reply src/client/splitpane.ts]: which also reminds me, why not derive the highlighting language from the file extension? 🙂 very low hanging fruit, boosts readability like crazy

File: src/client/splitpane.ts [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    return null;
+  }
+
+  function generateCode(client: LanguageClient, args: Array<string>) {
+    const currFile: string = window.activeTextEditor.document.fileName;
+    if (!(currFile.endsWith('.json') || currFile.endsWith('.md'))) {
+      window.showErrorMessage(currFile + ' is not a valid OpenAPI specification file format.');
+      return;
+    }
+    let autorestArgs = {};
```
Comment (olydis) [inline src/client/splitpane.ts]: add type please

**Review by olydis — COMMENTED** [review]
just some minor comments and ideas

File: src/client/splitpane.ts (hunk 2 above)
Comment (dsgouda) [reply src/client/splitpane.ts]: Makrdown allows language specific highlighting? Had no idea about it!

File: src/server/document-manager.ts [hunk 4]
```diff
@@ -547,6 +558,8 @@ export class OpenApiDocumentManager extends TextDocuments {
 
     connection.onDefinition((position, _cancel) => this.onDefinition(position));
 
+    connection.onRequest('autorest.generateCode', (args: any) => this.autoGenerateCode(args));
```
Comment (olydis) [inline src/server/document-manager.ts:563]: `args: AutorestArgs`

File: src/server/document-manager.ts [hunk 5]
```diff
@@ -531,6 +532,16 @@ export class OpenApiDocumentManager extends TextDocuments {
     ];
   }
 
+  private async autoGenerateCode(args: any): Promise<string> {
```
Comment (olydis) [inline src/server/document-manager.ts]: can take an `AutorestArgs` now 😉

File: src/lib/interfaces.ts [hunk 6]
```diff
@@ -17,4 +17,9 @@ export interface AutoRestSettings {
   debug: boolean;
   runtimeId: string;
   minimumAutoRestVersion: string;
+}
+
+export interface AutoRestCodeGenerationResult {
+  outputFolder: string;
+  generatedFiles: Map<string, string>;
```
Comment (olydis) [inline src/lib/interfaces.ts]: don't think you meant to use `Map`: it uses `set`/`get`/`has` for element access. I think you wanna have `{ [fileName: string]: string }`

File: src/server/document-context.ts [hunk 7]
```diff
@@ -53,16 +53,19 @@ export class DocumentContext extends EventEmitter implements IFileSystem {
     // where to output the generated code (clear the directory before generating code)
     const opPath: string = path.join(os.tmpdir(), opDirName);
     codeGeneratorInstance.AddConfiguration({ "output-folder": opPath });
-    result['outputFolder'] = opPath;
 
     this.Manager.listenForResults(codeGeneratorInstance);
-    let genFiles = {};
+    let genFiles = new Map<string, string>();
```
Comment (olydis) [inline src/server/document-context.ts]: same as above

File: src/lib/interfaces.ts (hunk 6 above)
Comment (dsgouda) [reply src/lib/interfaces.ts]: I needed that!

**Review by olydis — APPROVED** [review]
noiz

**Comment by dsgouda** [comment]
Closing since @fearthecowboy has included changes in a different PR

**Pull request closed without merging**